### DISCUSSION TITLE
Add Advanced feature to allow Form scripts

### DISCUSSION THREAD

**Discussion by emahuni** [directus/directus#19053] [Feature Requests]
### Summary

Add scripting support for advanced users to supply scripts that allow them to control forms and fields with access to stores, apis etc. These scripts can be served by the Directus api using a `/script` endpoint. This allows app extensibility to a whole new level.

There are times when you want to be able to control how fields behave, how they are displayed, filtered based on rules that can't be done with "conditions" configuration or calculations between fields on input. The idea is to have freedom of control over the entire form as a whole as a Vue engineer or just a js programmer.

Below is an example of scripts added to a "Quotas" collection. I added this to the end of the directus collections model configuration and allows the following configuration:

![image](https://github.com/directus/directus/assets/7290508/e60ddfde-d768-4a0f-884b-8be2f8e19095)

**Form setup script:** allows the api to fetch the configured script file from `extensions/scripts/configured_script_file.js`. This file can import other modules (not packages), just other js to allow modularisation of code and promote order. It is run when the form is setup and allows user to do something during that time.

**On Form field input script:** allows the api to fetch the configured script file from `extensions/scripts/configured_script_file.js`. File can also be structured as above. It is run on any input field, allowing for calculations or other meta data changes and behaviours based on field values.

More scripting opportunities can also be added such as `onBeforeSave`, and `onBeforeUnmount`.

I love how well though Directus is in accommodating non-programmers, which makes it really easy for advanced users. However, it limits advanced users from doing certain things as it doesn't provide any scripting. The best way to get around this is to fork and modify Directus APP, which in my opinion is not the best way to do it.

### Basic Example

 eg: being able to hide the "`create button`" of an "`o2m`" field if the form is not the first one; ie: nested relationships tend to stack forms as one enters information. Some forms relate to the same collection in many different ways but should be restricted in some way depending on the level of viewing or data entry. 

What I wanted to do was disable creation only if this isn't the root/first form in the stack so that a user can only add that field only from the original collection's item, not from a relationship. This **cannot** be achieved through conditions and to have such features requested all the time will blot the codebase. This is more of a custom use case, and should be able to do this if desired.

Below is the same collection form from the actual collections table:

![image](https://github.com/directus/directus/assets/7290508/3d6edd27-78fb-4d6a-93b0-502239075892)

![image](https://github.com/directus/directus/assets/7290508/008dbc9b-ba7c-4613-9992-6dc56aedb1b9)

if the user clicks that to view more details on the from inventory they get to see that:

![image](https://github.com/directus/directus/assets/7290508/0357f126-cea6-434e-9f94-c26cdaa9ee5b)

As you can see that is the 3rd form and is of the same collection as the first but points to a different item.
The `create `button for the same field is not showing because this is not the first form. I was able to hide that button through these scripts using logic that tests the above assumptions using a simple `if (formIndex !== 0) fields.quotas_in.meta.options.enableCreate = false;` line in the script.

### Motivation

There is no way to control the form to do things that can be done easily using js. There was no way of calculating fields or controlling how a form should behave without constantly modifying Directus App. I needed to just make minor modifications to the App and keep custom logic separate without the risk of using eval.

### Detailed Design

I have a working implementation of this already as I mentioned somewhere. The `/script` endpoint loads scripts from the server using `rollup` to ensure that any imports are bundled up into one script that is then sent to the browser as a response. Imports are not packages, just other script files in the same directory structure.

I modified `v-form` component to use a composable that loads these scripts and create functions that are executed at the appropriate moment with proper arguments passed to them.

ie: `form_setup` scripts are loaded during form setup and execute only once. This is just done in the setup function.

The `on_form_field_input_script` is executed whenever there is an input. This is invoked from `setValue` like so:
```js
  async function setValue(fieldKey: string, value: any, opts?: { force?: boolean }) {
	  const field = fieldsMap.value[fieldKey];
  
	  if (opts?.force !== true && (!field || isDisabled(field))) return;
  
    const edits = props.modelValue ? cloneDeep(props.modelValue) : {};
    edits[fieldKey] = value;
  
    if (runOnFormInputScript) await runOnFormInputScript(fieldKey, value, edits);
          
    emit('update:modelValue', edits);
  }
```

### Must Have:
- Ability to control each form independent of the other and differentiate each form.
- A stacked forms Pinia store allow forms to share data or know whether there are any forms before or after them.

### Drawbacks

There aren't any, really. The tiny modification I made is like a pass through. I made it to exist without impacting existing logic. 

One I can think of is that most advanced users will tend to use this more that existing conditions in the model configurations.

### Alternatives

I thought of extensions, but noticed that inorder to do this well I had to modify the v-form component. I have already done all this, and it works very well. Needed to use it. Hence I thought I could share this and ask the team to consider this for the core addition as it is so awesome. I have noticed I use it more often than not when designing the backoffice of the app I am making.

### Adoption Strategy

Documentation additions and modifications.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
